### ISSUE TITLE
problem with RAM allocating in FasterRCNN

### ISSUE THREAD

**Issue by buus2** [pytorch/vision#984]
Hello. First of all I'd like to thank for adding object detection models to torchvision, it's a great help for the community.

However, I encountered a problem while trying to use them. I just copied the example code from https://github.com/pytorch/vision/blob/3d5610391eaef38ae802ffe8b693ac17b13bd5d1/torchvision/models/detection/faster_rcnn.py#L102-L140
to a jupyter notebook and realized that during each execution of `model(x)` (on CPU) more than 2 GB of RAM is grabbed and not released afterwards. Running `del model` does not release RAM, only restarting the kernel does.

I met the same problem for the model defined in the following way:
```
model = torchvision.models.detection.fasterrcnn_resnet50_fpn(pretrained=True)
in_features = model.roi_heads.box_predictor.cls_score.in_features
model.roi_heads.box_predictor = FastRCNNPredictor(in_features, num_classes)
```
as stated in https://pytorch.org/tutorials/intermediate/torchvision_tutorial.html

What to do to get rid of this problem? Thanks in advance.

**Comment by fmassa** [comment]
The example model that is present in the documentation is not optimized at all for faster runtime, and indeed uses a lot of memory in the `rpn_head` because it has a huge convolution there (using 1280 input channels, and with 1280 output channels, for a large input!). You need to use a different `rpn_head` for it to be more memory efficient (for example, by having a conv going from 1280 channels to 128 or something like that).

The detection / instance segmentation models models in general use an image of minimum size of 800 pixels, and if you are running it in the CPU, it could dispatch to inefficient CPU kernels for the convolution, depending on how you installed PyTorch.

If you need to run it on smaller devices, try reducing the image size via `min_size` / `max_size` https://github.com/pytorch/vision/blob/3d5610391eaef38ae802ffe8b693ac17b13bd5d1/torchvision/models/detection/faster_rcnn.py#L57-L58

I'm closing this issue, but let me know if you still face the same problems.

**Comment by buus2** [comment]
Sorry for being unclear, I will try to explain my problem in a clearer way.

The huge size of the model is not a problem.
The problem is that each execution of `model(x)` takes a lot of RAM and does not release it afterwards. Therefore, RAM becomes clogged after several forwardpropagations.

**Comment by buus2** [comment]
The simplest example: notebook consisting of two cells:
```
import torch
import torchvision
model = torchvision.models.detection.fasterrcnn_resnet50_fpn(pretrained=True)
model.eval()
x = [torch.rand(3, 800, 800)] * 3
```
and
```
model(x)
```
After running the first cell the process takes 408 MB of RAM. Then after running `model(x)` several times it takes 2.9 GB, 5.1 GB, 7.2 GB, 9.5 GB and so on and it does not drop after 10 minutes. The outputs consist of empty tensors only (there are no bboxes predicted).

**Comment by fmassa** [comment]
@buus2 can you try
```python
with torch.no_grad():
    output = model(x)
```
and report back? There might be a reference that's kept in the forward that I might need to be fixed somewhere

**Comment by buus2** [comment]
@fmassa you are right, with `no_grad` RAM does not clog.
Could you please suggest a workaround for forwardpropagation during training?

**Comment by fmassa** [comment]
there should be no problem during training if you do `backward` afterwards.

What's your use-case? Computing the loss without updating the model parameters?

**Comment by fmassa** [comment]
Also, I couldn't reproduce the leak that you are mentioning.

The memory gets stabilized after the second forward, so there seems to be no references that are kept.

**Comment by buus2** [comment]
I am just trying to do an ordinary transfer learning. OK, so I will add `with torch.no_grad():` in the forwardpropagation in the training loop as well, thanks a lot!

**Comment by fmassa** [comment]
@buus2 as I mentioned before, just forwarding the model a number of times didn't show any leak for me, even without `torch.no_grad()`, so maybe you are holding references in your code to the output?
Note that if you do

```python
outputs = []
for i in range(10):
    outputs.append(model(input))
```
this will hold the full computational graph in memory, and will look like a memory leak.

I'm closing this as this is doesn't seem to be an issue with the model itself, but let me know if you still face issues

**Comment by buus2** [comment]
I experience the described issue in the notebook consisting of exactly the 6 lines of code that I wrote above.

I use:
- Python 3.6.7 (default, Oct 22 2018, 11:32:17) \n[GCC 8.2.0]
- Jupyter Lab 0.34.8
- PyTorch 1.1.0
- TorchVision 0.3.0

I will experiment with other Python and Jupyter Notebook/Lab versions. If I find something interesting, I will report it here. Thanks for help!

**Comment by fmassa** [comment]
@buus2 maybe when you do `model(x)` in a jupyter notebook it holds the data internally?

Could you try doing
```python
o = model(x)
print(o)
```
instead and see if it still leaks?

**Comment by apyskir** [comment]
Hi, I observe similiar behaviour as @buus2 observes. Each time I do:

```
o = model(x)
print(o)
```

my free RAM decreases by around 0.2 GB and each time I do:
```
model(x)
```
my free RAM decreases by around 2 GB. Looping:
```
for _ in range(100):
  model(x)
```
quickly leads to memory error.

Python 2.7.15
jupyter 4.4.0
pytorch 1.1.0
torchvision 0.3.0

**Comment by buus2** [comment]
I switched to a bare python script and used [memory_profiler](https://pypi.org/project/memory-profiler/). This is what I got:
```
Line #    Mem usage    Increment   Line Contents
================================================
    10  408.156 MiB  408.156 MiB   @profile
    11                             def run():
    12 1846.680 MiB 1438.523 MiB       model(x)
    13 3385.992 MiB 1539.312 MiB       o = model(x)
    14 3385.992 MiB    0.000 MiB       time.sleep(60)
    15 4518.891 MiB 1132.898 MiB       model(x)
    16 4989.570 MiB  470.680 MiB       o = model(x)
    17 4989.570 MiB    0.000 MiB       time.sleep(60)
    18 5520.141 MiB  530.570 MiB       model(x)
    19 5934.988 MiB  414.848 MiB       o = model(x)
    20 5934.988 MiB    0.000 MiB       time.sleep(60)
    21 5934.988 MiB    0.000 MiB       with torch.no_grad():
    22 5935.156 MiB    0.168 MiB           model(x)
    23 5205.539 MiB    0.000 MiB           o = model(x)
    24 5205.539 MiB    0.000 MiB           time.sleep(60)
    25 5205.566 MiB    0.027 MiB           model(x)
    26 5205.594 MiB    0.027 MiB           o = model(x)
    27 5205.594 MiB    0.000 MiB           time.sleep(60)
    28 5205.621 MiB    0.027 MiB           model(x)
    29 5205.504 MiB    0.000 MiB           o = model(x)
    30 5205.504 MiB    0.000 MiB           time.sleep(60)
```
I tried several times, the results were consistent.
Well, I have no idea, I will just put `with torch.no_grad()` everywhere.

**Comment by fmassa** [comment]
@buus2 @apyskir I digged a bit in this, and it's a known behavior of jupyter.
See this comment for example https://github.com/pyviz/holoviews/issues/1821#issuecomment-324941357

> Here the second cell was re-executed a bunch of times but the notebook keeps holds onto every intermediate output, even if the user doesn't define their own handle on it!

Which means that jupyter is actually keeping in memory **all** the intermediate outputs from your cells.
When you do `with torch.no_grad()`, you are just saving memory because you don't keep the full computation graph in memory, but just the outputs (which are very small). But the "holding" into memory from jupyter still happens.

Given that this is a known behavior from jupyter, I don't think there is anything we can do in torchvision about it.

**Comment by buus2** [comment]
@fmassa Please take a look at my comment above.

**Comment by fmassa** [comment]
@buus2 hum, that's interesting. There is definitely a "leak" due to jupyter, but there also seems to be something weird happening with our CPU backend when not set to `no_grad()` mode.

I've just tried the exact same snippet with `resnet50` instead of `fasterrcnn_resnet50_fpn`, and I also experienced this memory increase on the CPU (everything stays constant on the GPU), so this is not a problem specific to `fasterrcnn`.

My guess would be that we don't free some references in the MKLDNN backend for the convolutions, but I should investigate a bit more.

@buus2 in the meantime, could you double-check that running with `resnet50` on the CPU also leads to an apparent leak?

**Comment by fmassa** [comment]
For completeness, here is my script

```python
import torch, torchvision


@profile
def run():
    model = torchvision.models.resnet50(pretrained=True).eval()
    x = torch.rand(32, 3, 224, 224)
    model(x)
    model(x)
    model(x)
    model(x)
    model(x)
    model(x)
    model(x)
    model(x)
    model(x)
    model(x)

if __name__ == "__main__":
    run()
```
 and the result of using `memory_profile`
```
Line #    Mem usage    Increment   Line Contents
================================================
     4  169.457 MiB  169.457 MiB   @profile
     5                             def run():
     6  371.051 MiB  201.594 MiB       model = torchvision.models.resnet50(pretrained=True).eval()
     7  389.418 MiB   18.367 MiB       x = torch.rand(32, 3, 224, 224)
     8 2460.738 MiB 2071.320 MiB       model(x)
     9 2950.367 MiB  489.629 MiB       model(x)
    10 3507.016 MiB  556.648 MiB       model(x)
    11 3972.516 MiB  465.500 MiB       model(x)
    12 4511.520 MiB  539.004 MiB       model(x)
    13 4952.520 MiB  441.000 MiB       model(x)
    14 5148.512 MiB  195.992 MiB       model(x)
    15 5565.020 MiB  416.508 MiB       model(x)
    16 5810.012 MiB  244.992 MiB       model(x)
    17 5834.516 MiB   24.504 MiB       model(x)
```

**Comment by buus2** [comment]
For
```
model = torchvision.models.resnet50(pretrained=True).eval()
x = torch.rand(3, 3, 800, 800)
```
I obtain
```
Line #    Mem usage    Increment   Line Contents
================================================
    12  290.652 MiB  290.652 MiB   @profile
    13                             def run():
    14 1630.414 MiB 1339.762 MiB       model(x)
    15 3608.125 MiB 1977.711 MiB       o = model(x)
    16 3608.125 MiB    0.000 MiB       time.sleep(60)
    17 4939.422 MiB 1331.297 MiB       model(x)
    18 5437.441 MiB  498.020 MiB       o = model(x)
    19 5437.441 MiB    0.000 MiB       time.sleep(60)
    20 5818.457 MiB  381.016 MiB       model(x)
    21 6023.277 MiB  204.820 MiB       o = model(x)
    22 6023.277 MiB    0.000 MiB       time.sleep(60)
    23 6023.277 MiB    0.000 MiB       with torch.no_grad():
    24 6023.277 MiB    0.000 MiB           model(x)
    25 4910.070 MiB    0.000 MiB           o = model(x)
    26 4910.070 MiB    0.000 MiB           time.sleep(60)
    27 4910.070 MiB    0.000 MiB           model(x)
    28 4910.070 MiB    0.000 MiB           o = model(x)
    29 4910.070 MiB    0.000 MiB           time.sleep(60)
    30 4910.070 MiB    0.000 MiB           model(x)
    31 4910.070 MiB    0.000 MiB           o = model(x)
    32 4910.070 MiB    0.000 MiB           time.sleep(60)
```
So probably you are right, there is a problem with resnet. Please also note the usage drop in line 25 (and the analogous one in line 23 for FasterRCNN above).

**Comment by fmassa** [comment]
Ok, I think I've found the problem (and the solution).

Funnily, it has already happened to me 4 years ago in https://github.com/torch/torch7/issues/229, and the solution in that thread https://github.com/torch/torch7/issues/229#issuecomment-102870888 applies here.

> "malloc'ed memory is not always released back to the OS, especially for small object sizes. (Larger allocations will use mmap() / munmap() directly.) The memory can be reused in the same process, though." - Tudor
> 
> "There are better memory allocators out there, such as our own Jason Evans's jemalloc", "(jemalloc does make a good faith effort to release memory back to the OS if possible; shrinking the data segment is hard because you can't defragment without a GC, so jemalloc uses mmap() rather than sbrk())".

So, if we use jemalloc instead of the default malloc, things should work fine.

Here is an example with malloc:

```
(segmentation) fmassa@devfair0163:~/work/video_loader$ python -m memory_profiler leak.py
Filename: leak.py

Line #    Mem usage    Increment   Line Contents
================================================
     4  169.598 MiB  169.598 MiB   @profile
     5                             def run():
     6  371.188 MiB  201.590 MiB       model = torchvision.models.resnet50(pretrained=True).eval()
     7  389.578 MiB   18.391 MiB       x = torch.rand(32, 3, 224, 224)
     8 2499.969 MiB 2110.391 MiB       model(x)
     9 2994.383 MiB  494.414 MiB       model(x)
    10 3484.387 MiB  490.004 MiB       model(x)
    11 3925.387 MiB  441.000 MiB       model(x)
    12 4292.887 MiB  367.500 MiB       model(x)
    13 4611.383 MiB  318.496 MiB       model(x)
    14 4856.387 MiB  245.004 MiB       model(x)
    15 5101.383 MiB  244.996 MiB       model(x)
    16 5370.887 MiB  269.504 MiB       model(x)
    17 5319.277 MiB    0.000 MiB       model(x)
```

and we see that the memory seems to increase over time. Now let's use `jemalloc`:
```
(segmentation) fmassa@devfair0163:~/work/video_loader$ LD_PRELOAD=/usr/lib/x86_64-linux-gnu/libjemalloc.so.1 python -m memory_profiler leak.py
Filename: leak.py

Line #    Mem usage    Increment   Line Contents
================================================
     4  171.730 MiB  171.730 MiB   @profile
     5                             def run():
     6  374.852 MiB  203.121 MiB       model = torchvision.models.resnet50(pretrained=True).eval()
     7  393.215 MiB   18.363 MiB       x = torch.rand(32, 3, 224, 224)
     8 1222.484 MiB  829.270 MiB       model(x)
     9 1227.727 MiB    5.242 MiB       model(x)
    10 1226.277 MiB    0.000 MiB       model(x)
    11 1230.836 MiB    4.559 MiB       model(x)
    12 1225.773 MiB    0.000 MiB       model(x)
    13 1231.008 MiB    5.234 MiB       model(x)
    14 1210.758 MiB    0.000 MiB       model(x)
    15 1230.848 MiB   20.090 MiB       model(x)
    16 1215.566 MiB    0.000 MiB       model(x)
    17 1213.285 MiB    0.000 MiB       model(x)
```

And now we see that the memory stays nicely the same.

TL;DR: this is not a leak in PyTorch nor torchvision, but instead a known (and unintuitive) behavior of malloc.

**Comment by apyskir** [comment]
Hi, I obtained very same results as @buus2 but while running the script on GCP VM instance. After looking at your last post @fmassa it's not surprising that the problem arises everywhere. Thanks

**Comment by buus2** [comment]
@fmassa I am glad you finally found it. Could you please tell us what we should do now? Do we have to remember about indicating the right memory allocator in C each time we want to run inference on CPU?

**Comment by fmassa** [comment]
@buus2 this is not a leak, and you shouldn't face OOM errors because of that.

As workarounds, use `torch.no_grad()`, and maybe use `jemalloc` when running your programs.
